### PR TITLE
5298 seperate keycloak clients

### PR DESCRIPTION
**Pull request by tschia**
also fixed some testcafe tests

### CODE REVIEW

**Review by weskubo-cgi — CHANGES_REQUESTED** [review]

File: backend/src/main/java/ca/bc/gov/hlth/hnweb/security/AudienceValidator.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import org.springframework.security.oauth2.core.OAuth2TokenValidatorResult;
+import org.springframework.security.oauth2.jwt.Jwt;
+import org.springframework.stereotype.Component;
+import org.springframework.util.Assert;
+
+import static org.springframework.security.oauth2.jwt.JwtClaimNames.AUD;
+
+/*
+ The generic Spring JwtClaimValidator throws an NPE when a claim is missing, so we customize it to avoid that exception
+ */
```
[the remark this answers is not in the document]
Comment (tschia) [reply backend/src/main/java/ca/bc/gov/hlth/hnweb/security/AudienceValidator.java]: fixed

File: backend/src/main/java/ca/bc/gov/hlth/hnweb/security/AudienceValidator.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+@Component
+public class AudienceValidator implements OAuth2TokenValidator<Jwt> {
+
+    static final Logger logger = LoggerFactory.getLogger(SecurityConfig.class);
```
[the remark this answers is not in the document]
Comment (tschia) [reply backend/src/main/java/ca/bc/gov/hlth/hnweb/security/AudienceValidator.java]: fixed

File: backend/src/main/java/ca/bc/gov/hlth/hnweb/security/AudienceValidator.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+    @Value("${spring.security.oauth2.resourceserver.jwt.audience}")
+    private String audience;
+
+    private final OAuth2Error error = new OAuth2Error(OAuth2ErrorCodes.INVALID_REQUEST, "The " + AUD + " claim is not valid",
```
[the remark this answers is not in the document]
Comment (tschia) [reply backend/src/main/java/ca/bc/gov/hlth/hnweb/security/AudienceValidator.java]: fixed

File: backend/src/main/java/ca/bc/gov/hlth/hnweb/security/SecurityConfig.java [hunk 4]
```diff
@@ -3,6 +3,8 @@
 import java.util.Arrays;
 import java.util.Collections;
 
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
```
[the remark this answers is not in the document]
Comment (tschia) [reply backend/src/main/java/ca/bc/gov/hlth/hnweb/security/SecurityConfig.java]: fixed

File: backend/src/main/java/ca/bc/gov/hlth/hnweb/security/SecurityConfig.java [hunk 5]
```diff
@@ -12,22 +14,33 @@
 import org.springframework.security.config.annotation.web.builders.HttpSecurity;
 import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
 import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter;
+import org.springframework.security.oauth2.core.*;
+import org.springframework.security.oauth2.jwt.*;
 import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationConverter;
+import org.springframework.util.Assert;
```
[the remark this answers is not in the document]
Comment (tschia) [reply backend/src/main/java/ca/bc/gov/hlth/hnweb/security/SecurityConfig.java]: fixed

File: backend/src/main/java/ca/bc/gov/hlth/hnweb/security/SecurityConfig.java [hunk 6]
```diff
@@ -12,22 +14,33 @@
...
 import org.springframework.web.cors.CorsConfiguration;
 import org.springframework.web.cors.CorsConfigurationSource;
 import org.springframework.web.cors.UrlBasedCorsConfigurationSource;
 
+import static org.springframework.security.oauth2.jwt.JwtClaimNames.AUD;
```
[the remark this answers is not in the document]
Comment (tschia) [reply backend/src/main/java/ca/bc/gov/hlth/hnweb/security/SecurityConfig.java]: fixed

File: backend/src/main/java/ca/bc/gov/hlth/hnweb/security/AudienceValidator.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+/*
+ The generic Spring JwtClaimValidator throws an NPE when a claim is missing, so we customize it to avoid that exception
+ */
+@Component
+public class AudienceValidator implements OAuth2TokenValidator<Jwt> {
+
+    static final Logger logger = LoggerFactory.getLogger(SecurityConfig.class);
+
+    @Value("${spring.security.oauth2.resourceserver.jwt.audience}")
```
[the remark this answers is not in the document]
Comment (tschia) [reply backend/src/main/java/ca/bc/gov/hlth/hnweb/security/AudienceValidator.java:24]: as discussed offline, left as is for now

**Review by weskubo-cgi — APPROVED** [review]

**Pull request merged**